wakuv2/waku.go

```diff
@@ -1039,38 +944,202 @@ func (w *Waku) Subscribe(f *common.Filter) (string, error) {
-		if _, err := w.node.FilterLightnode().Unsubscribe(ctx, contentFilter); err != nil {
-			return fmt.Errorf("failed to unsubscribe: %w", err)
-		}
-	}
-
 	ok := w.filters.Uninstall(id)
 	if !ok {
 		return fmt.Errorf("failed to unsubscribe: invalid ID '%s'", id)
 	}
+
 	return nil
 }
 
+func (w *Waku) runFilterLoop() {
+	defer w.wg.Done()
+
+	if !w.settings.LightClient {
+		return
+	}
+
+	// Use it to ping filter peer(s) periodically
+	ticker := time.NewTicker(time.Duration(w.cfg.KeepAliveInterval) * time.Second)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-w.ctx.Done():
+			return
+		case <-ticker.C:
+			w.checkFilterHealth()
+		}
+	}
+}
+func (w *Waku) runFilterSubscriptionLoop(sub *filter.SubscriptionDetails) {
+	for {
+		select {
+		case <-w.ctx.Done():
+			return
+		case env, ok := <-sub.C:
+			if ok {
+				err := w.OnNewEnvelopes(env, common.RelayedMessageType)
+				if err != nil {
+					w.logger.Error("OnNewEnvelopes error", zap.Error(err))
+				}
+			} else {
+				return
+			}
+		}
+	}
+}
+
+// Initial subscribe handler
+func (w *Waku) filterSubscribeHandler(f *common.Filter) {
+	if !w.settings.LightClient {
+		return
+	}
+	peers := w.findFilterPeers()
+	if len(peers) > 0 {
+		contentFilter := w.buildContentFilter(f.PubsubTopic, f.ContentTopics)
+		// Take first available filter peer
+		w.subscribeToFilter(peers[0], contentFilter)
+	} else {
+		w.logger.Error("could not select a suitable peer for filter")
+	}
+}
+
+func (w *Waku) filterUnsubscribeHandler(f *common.Filter) {
+	if !w.settings.LightClient {
+		return
+	}
+	contentFilter := w.buildContentFilter(f.PubsubTopic, f.ContentTopics)
+	if _, err := w.node.FilterLightnode().Unsubscribe(w.ctx, contentFilter); err != nil {
+		w.logger.Error("failed to unsubscribe: %w", zap.Error(err))
+	}
+}
+
+func (w *Waku) findSubsForFilter(f *common.Filter) filter.SubscriptionSet {
+
+	// TODO improve performance
+	subscriptions := w.node.FilterLightnode().Subscriptions()
+	subsForFilter := make(filter.SubscriptionSet)
+	contentFilter := w.buildContentFilter(f.PubsubTopic, f.ContentTopics)
+	for _, sub := range subscriptions {
+		if sub.ContentFilter.PubsubTopic == contentFilter.PubsubTopic &&
+			maps.Equal(sub.ContentFilter.ContentTopics, contentFilter.ContentTopics) {
+			subsForFilter[sub.ID] = sub
+		}
+	}
+
+	return subsForFilter
+}
+
+// Check whether each of the installed filters
+// has enough alive subscriptions to peers
+func (w *Waku) checkFilterHealth() {
+
+	w.logger.Info("In checkFilterHealth()")
+	w.filters.IterateFilters(func(filterId string) {
+		f := w.filters.Get(filterId)
+		subsForFilter := w.findSubsForFilter(f)
+		w.logger.Info("Checking filter", zap.String("filterId", filterId), zap.Any("len(subsForFilter)", len(subsForFilter)))
+		activeSubs := make(filter.SubscriptionSet)
+
+		for subID, sub := range subsForFilter {
+			w.logger.Info("Check sub aliveness", zap.Stringer("peerId", sub.PeerID), zap.Strings("topics", maps.Keys(sub.ContentFilter.ContentTopics)))
```

Comment:
Wouldn't this add too much noise to the logs? Perhaps it should be `Debug`

Reply:
Yes, changed to Debug.